include/antler/project/project.hpp

```diff
@@ -49,132 +63,162 @@ class project {
-
-   /// Populate the directory by generating files.
-   /// @param replace  The should replace during population.
-   /// @param error_stream  The stream to print failure reports to.
-   /// @return true for success; false for failure.
-   [[nodiscard]] bool populate(bool replace = true, std::ostream& error_stream = std::cerr) noexcept;
-
-
-   /// Factory function.
-   /// @note The returned project may not be valid. The only guarantee is that parsing did not fail.
-   /// @note see is_valid() to test validity.
-   /// @param path  The location of the project.yaml file or the path containing it.
-   /// @param error_stream  The stream to print failure reports to.
-   /// @return std::optional containing a project if parsing succeeded.
-   [[nodiscard]] static std::optional<project> parse(const std::filesystem::path& path, std::ostream& error_stream = std::cerr);
+   [[nodiscard]] bool has_valid_dependencies() const noexcept;
 
    /// Initialize the directories
    /// @param path  The location of the project.yaml file or the path containing it.
-   /// @param error_stream  The stream to print failure reports to.
-   /// @return true for success; false indidates failure.
-   [[nodiscard]] static bool init_dirs(const std::filesystem::path& path, std::ostream& error_stream = std::cerr) noexcept;
+   /// @return true for success; false indicates failure.
+   [[nodiscard]] static bool init_dirs(const system::fs::path& path) noexcept;
 
    /// Search this and directories above for `project.yaml` file.
    /// @note if path extension is `.yaml` no directory search is performed, instead return value indicating existence of path a regular file.
    /// @param path  This is the search path to begin with; if the project file was found, it is updated to the path to that file.
    /// @return true if the project file was found and is a regular file; otherwise, false.
-   [[nodiscard]] static bool update_path(std::filesystem::path& path) noexcept;
+   [[nodiscard]] static bool update_path(system::fs::path& path) noexcept;
+
+   /// Serialization function from version to yaml node
+   [[nodiscard]] inline yaml::node_t to_yaml() const noexcept { 
+      yaml::node_t node;
+      node["project"] = m_name;
+      node["version"] = m_ver;
+      for (const auto& [k,v] : apps())
+         node["apps"].push_back(v);
+      for (const auto& [k,v] : libs())
+         node["libs"].push_back(v);
+      return node; 
+   }
 
-private:
-   inline void create_app(std::string_view name) {
+   /// Deserialization function from yaml node to version
+   [[nodiscard]] inline bool from_yaml(const yaml::node_t& n) noexcept {
+      return ANTLER_EXPECT_YAML(n, "project", name, std::string) &&
+             ANTLER_EXPECT_YAML(n, "version", version, antler::project::version) &&
+             ANTLER_TRY_YAML_ALL(n, "apps", upsert, antler::project::app_t) &&
+             ANTLER_TRY_YAML_ALL(n, "libs", upsert, antler::project::lib_t);
```

Comment:
This is short-circuit code and the logger will only report the first missing label. Would it be better to do this as a series of `&=` calls?